Paper: Enhancing Curve Fitting with SciPy: Interactive Spline Modeling and Reproducibility with SplineCloud

nomad-vagabond opened the pull request:
This article explores how SciPy’s spline fitting tools power the core of SplineCloud — an open platform for interactive constructing and sharing regression models. It introduces a modern workflow for creating, refining, and reusing curve-fitting models, making data modeling and analysis more transparent, collaborative, and reproducible.

Note: this is a draft paper, completed by 10%. However, it should already give an understanding of the subject. For reference, please take a look at these blog post that shares some of the concepts:

[Approximation by Spline Functions and Parametric Spline Curves with SciPy](http://vadym-pasko.com/blog/2015/03/06/spline-approx-scipy.html)

[Mastering Curve Fitting. How to Fit Complex Data with Parametric Splines](https://splinecloud.com/blog/mastering-curve-fitting-how-to-fit-complex-data-with-parametric-splines/)

nomad-vagabond committed:
Add initial draft of the Vadym Pasko paper

nomad-vagabond committed:
Update section 2.Spline Fitting

nomad-vagabond committed:
Add basic examples of spline fitting with SciPy

nomad-vagabond committed:
Add section 3.3 Pain Points of Pure Programmable Spline Fitting

apaleyes commented:
Hello! I've been assigned as a reviewer to this paper. It is clear the paper is currently being actively worked, so perhaps best thing to do for me now is to passively observe, and at some point @nomad-vagabond you could ping me when it's ready for review. Does that sound like a reasonable plan?

nomad-vagabond commented:
@apaleyes Sure! I'm planning to finish it over the coming weekend. I'll let you know once it is done.

nomad-vagabond committed:
Add section 4

nomad-vagabond committed:
Add draft for section 5

ameyxd commented:
Inviting reviewer: @ajinkyakadam

nomad-vagabond committed:
Add section 5.2 Automatic Spline Fitting

nomad-vagabond committed:
Add section draft: 5.3. Fine-Tuning. Interactive Adjustments of Control Points and Knot Vector

nomad-vagabond committed:
Finalize section 5.3. Fine-Tuning. Interactive Adjustments of Control Points and Knot Vector

nomad-vagabond committed:
Add draft of the subsection 5.4 Fitting Errors.

nomad-vagabond committed:
Finalize section 5

nomad-vagabond committed:
Add section 6. Reusability and Reproducibility with SplineCloud

apaleyes commented:
@nomad-vagabond just wanted to check if it's okay to start reviewing at least parts of the paper?

nomad-vagabond commented:
> @nomad-vagabond just wanted to check if it's okay to start reviewing at least parts of the paper?

@apaleyes  please, give me one more day to polish the main part, so you do not waste your attention on typos and other artifacts.  I will leave a message here once it is ready for initial review.

nomad-vagabond committed:
Add LSQUnivariateSpline usage example. Refactor and adjust style.

nomad-vagabond commented:
@apaleyes You can start a review process. I fixed main style issues and refactored figures and formulas. I'll be working on the last two sections in the meantime.

nomad-vagabond committed:
Use references for Figures and Formulas

nomad-vagabond committed:
Clean up rudiments from the template

apaleyes commented:
Thanks, @nomad-vagabond ! I skimmed through the paper, and have some early questions, before diving into details.

1. The paper is already above both the word (7k+>6k) and page (24>8) limit. It probably makes no difference for the HTML version, but matters more for PDFs. Of course there are some small cuts that can be done (e.g notebook-like code snippets in Section 3 can be greatly reduced), but these won't be nearly enough. I therefore wonder if the best approach would be to split it into main text and extended supplementary material. Or perhaps you had other thoughts?

2. It is mentioned multiple times that SplineCloud is an open platform, yet it is not clear what exactly does this mean. Is it free to use for everyone? Are there any free-tier limits? Can users run it locally (not just the client), or is it SaaS only on `splinecloud.com`? Whatever is the monetisation approach here is not important, but what is important is to be clear to which extent the platform is available to the users.

nomad-vagabond commented:
@apaleyes Thank you for the early comments. Somehow, I missed the requirements for the paper volume. I will shrink the text and maybe shorten or move code snippets out of the paper body. 

SplineCloud is free to use for everyone. It is SaaS. As of now, limits are only per repo size (under 3Gb). Users can not run the platform locally. I will add this information to the platform description.

nomad-vagabond committed:
Shorten sections 2 and 3. Add notebook with basic spline fitting examples using SciPy

ameyxd commented:
@cdlindsey will serve as editor for this paper.

nomad-vagabond committed:
Shorten and refactor sections 1-3

nomad-vagabond committed:
Reduce number of images

nomad-vagabond committed:
Replace grid statements with HTML tables for images alignment to avoid PDF build errors

nomad-vagabond committed:
Reduce text in some sections

nomad-vagabond committed:
More text reduction

fwkoch commented on papers/vadym_pasko/myst.yml:
```diff
@@ -0,0 +1,43 @@
+        - terradesert
+        - jupyter
+        - sklearn1
+        - sklearn2
```
@nomad-vagabond - Thanks for the submission! Your remaining automated check failures are related to citations with no DOI listed. Please see if your references have DOIs and add them if possible. If they do not have DOIs, you may add the keys to the list here and the checks will skip them. Something like:

```suggestion
  error_rules:
    - rule: doi-exists
      severity: ignore
      keys:
        - pasko-blog-post-2015,
```

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+## Some Theoretical Background Behind Splines
+
+Splines are piecewise-defined functions used extensively in numerical analysis, computer-aided geometric design, and data fitting. The fundamental idea behind spline fitting process is to construct a smooth function that matches a set of data points or satisfies a set of constraints, while preserving computational efficiency and numerical stability. Splines can be represented in several forms, ea ...
```
Amogst -> Amongst

apaleyes commented on papers/vadym_pasko/main.md:
````diff
@@ -0,0 +1,476 @@
+:label: b-spline
+\mathbf{C}(t) = \sum_{i=0}^{n} \mathbf{P}_i B_{i,k}(t)
+```
+where $\mathbf{P}_i$ are control points, $B_{i,k}(t)$ are the B-spline basis functions of degree $k$, defined recursively by the Cox–de Boor formula [@doi:10.1093/imamat/10.2.134; @doi:10.1007/978-1-4612-6333-3]:
````
second citation link does not work for me, goes to "page not found" in springer

nomad-vagabond replied:
fixed. added proper item to bibliography

apaleyes commented on papers/vadym_pasko/main.md:
````diff
@@ -0,0 +1,476 @@
+```
+where $\mathbf{P}_i$ are control points, $B_{i,k}(t)$ are the B-spline basis functions of degree $k$, defined recursively by the Cox–de Boor formula [@doi:10.1093/imamat/10.2.134; @doi:10.1007/978-1-4612-6333-3]:
+
+```{math}
````
Just wondering if B_i,0 should be defined here as a base case of recursion

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+### Applications and Benefits of Parametric Splines
+
+Parametric splines are widely used in computer graphics, computer-aided desig (CAD), and more general geometric modeling, where the goal is to design and manipulate accurate and smooth curves and surfaces in two or three dimensions. Their unique properties allow for high shape flexibility, local control over curve shape and its continuity. These properties make parametric splines ideal for applic ...
```
desig_n_

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+An intuitive visual explanation of splines, different forms of their representation and unique properties is given in the video by Freya Holmér [@holmer-continuity-of-splines].
+
+Despite their geometric advantages, parametric splines are rarely used in statistical data fitting and regression analysis. The primary limitation is the parameterization problem: for observed data points $(x_i, y_i)$, there is no natural way to assign parameter values $t_i$ that correspond meaningfully to the underlying data relationship. Additionally, parametric splines do not provide the direc ...
```
"there is no natural way" would be nice to support this strong statement with a citation

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+In {ref}`cox-de-boor` possible divisions by zero are resolved by the convention that ‘anything divided by zero is zero’ [@lyche-morken]. This formulation separates geometry (control points) from basis functions, allowing flexible manipulation and efficient computation of spline curves.
+
+### Applications and Benefits of Parametric Splines
```
I felt this subsection was not critically important for the paper. Feel free to keep it of course, but if editors ask to reduce paper size further this might be an easy candidate for deletion

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+### Pain Points of Pure Programmable Spline Fitting
+
+SciPy provides a programmatic interface to robust spline fitting methods suitable for a variety of tasks, particularly when working with complex experimental or statistical data. However, despite offering multiple fitting methods and adjustable parameters, selecting an appropriate combination can be challenging. As noted in [@pasko-blog-post-2015], issues such as overfitting and extrapolation con ...
```
Would look a bit nicer if written as `As noted by @pasko-blog-post-2015`

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+SplineCloud is an open platform for formalized knowledge exchange. It is designed to make technical data, like results of simulations, tests and modeling more accessible and reusable in computer code. Its architecture and instruments help reduce the routine overhead of data collection, processing, and model reconstruction in code.
+
+SplineCloud is free to use for everyone, publishing open data, however its source code is closed (as of now). Repositories with limited access is a part of the platform monetization model. All tools are free to use without limitaions (excpt repository size, that as of now is limited to 3 Gb). The platform's main features are: plot digitizer, advanced spline fitting tool, open API, and client libr ...
```
limita_t_ions, exc_e_pt

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+:::{figure} workflow_scheme.png
+:label: fig:workflow_scheme
+:width: 400px
+SplineCloud workflow scheme
```
Consider expanding this caption, perhaps describing the user journey with text.

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+SplineCloud objects relationship diagram
+:::
+
+Spline fitting tool, a main instrument of the platform, is built on top of SciPy’s interpolate module and provides a visual interface to three main spline fitting methods described in section 3: Interpolating Splines, Smoothing Splines, Least Squares Splines.
```
"described in section 3" - this should probably be a hyperlink

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+Spline fitting tool
+:::
+
+The important difference here is that SplineCloud’s smoothness parameter is a relative parameter used to calculate the actual SciPy’s (FITPACK’s) smoothing factor $s$ defined as:
```
"The important difference" difference with what?

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+### Fine-Tuning. Interactive Adjustments of Control Points and Knot Vector
+
+A principal enhancement over the automatic fitting approach is SplineCloud’s *Fine-Tuning* functionality. It provides the ability to visually adjust control points and knot vectors of fitted splines. As proved in many cases, this interface enables users to achieve curve refinements that exceed the capabilities of SciPy’s automatic fitting algorithms in terms of smoothness and accuracy. More of th ...
```
More of that -> Moreover

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+An official Python client, `splinecloud-scipy`, is based on SciPy and allows for fetching spline data using the spline UIDs and recreating the model in code. Reusable spline model has critical methods for evaluating spline in the form $y=f(x),$ loading underlying data, and assessing fit accuracy (using one of the methods, listed in Section 5.4). This enables a new level of interoperability: a spl ...
+
+Let’s first take a look at the client library basic usage scenarios and then discuss the library structure and how splines a recreated and evaluated. 
```
splines a -> splines are

apaleyes commented on papers/vadym_pasko/main.md:
````diff
@@ -0,0 +1,476 @@
+ self.spline_x = si.UnivariateSpline._from_tck(tck_x)
+ self.spline_y = si.UnivariateSpline._from_tck(tck_y)
+```
+There is a plan to use the `BSpline` class instead.
````
This sentence felt a bit out of place

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,476 @@
+
+The interactive, collaborative and reusable curve fitting enabled by SplineCloud minimizes redundant work and enhances the reliability and continuity of model-based research outputs. However, the platform currently does not support other classes of regression models such as polynomial fits, rational functions, exponential models, or machine learning-based regressors. This limitation is not techni ...
+
+Spline models, particularly B-splines and their parametric forms, possess a standardized and well-defined mathematical representation. This makes them ideally suited for platform-independent storage, manipulation, and code-level reuse. In contrast, arbitrary analytical or statistical regression models lack the generalization inherent to splines and the support of all possible modeling techniqus i ...
```
techniqus -> techniques

apaleyes reviewed (commented):
This is a very well written paper that gives a very brief overview  of splines followed by an in-depth presentation of SplineCloud (with some comparisons with SciPy along the way). The narrative is very clear and logical, overview of functionality appears to be quite comprehensive, pros and cons of using SplineCloud vs SciPy are clear. Thanks!

The unfinished section of use cases and applications is highly recommended, showing how SplineCloud can be useful in real life will make the paper much more impactful and impressive, much more so than detailed description of the user interface.

Additional question is about SplineCloud itself. It is quite often that engineers end up using markup/config as code even with tools that provide GUI editing. I have an impression SplineCloud is GUI only, is that correct? Any plans to allow power usage via code/configs?

Separately, it is important to keep in mind that this paper describes a free (for now) functionality of a pay-for platform with closed source (all of this is clearly stated in the text). The paper right now still renders into 15 pages PDF,  almost 2x the stated limit. Personally I have not problem with either point, it's more up to the editors how strict they want to be.

nomad-vagabond committed:
Add first use case to Section 7

nomad-vagabond committed:
Add exceptions to the reference without DOIs

nomad-vagabond committed:
Add second use case to Section 7

nomad-vagabond committed:
Add third use case to Section 7

nomad-vagabond committed:
Add Section 8 Summary and Future Directions

nomad-vagabond commented:
@apaleyes Thank you for the review and feedback!  I will incorporate your comments and suggestions later today.

I added missing sections and they are ready for review.

nomad-vagabond committed:
Address comments int implement suggestions from reviewers

nomad-vagabond commented:
@apaleyes, all your comments and suggestions have been implemented. There may be some minor improvements after spelling check, and others.

nomad-vagabond commented:
> This is a very well written paper that gives a very brief overview of splines followed by an in-depth presentation of SplineCloud (with some comparisons with SciPy along the way). The narrative is very clear and logical, overview of functionality appears to be quite comprehensive, pros and cons of using SplineCloud vs SciPy are clear. Thanks!
> 
> The unfinished section of use cases and applications is highly recommended, showing how SplineCloud can be useful in real life will make the paper much more impactful and impressive, much more so than detailed description of the user interface.
> 
> Additional question is about SplineCloud itself. It is quite often that engineers end up using markup/config as code even with tools that provide GUI editing. I have an impression SplineCloud is GUI only, is that correct? Any plans to allow power usage via code/configs?
> 
> Separately, it is important to keep in mind that this paper describes a free (for now) functionality of a pay-for platform with closed source (all of this is clearly stated in the text). The paper right now still renders into 15 pages PDF, almost 2x the stated limit. Personally, I have no problem with either point, it's more up to the editors how strict they want to be.

@apaleyes Thank you for the feedback. Answering your questions:
- SplineCloud is GUI only, but it is intended to facilitate the development of model-based algorithms by exposing all objects over the API. Such areas as MDO (Multidisciplinary design optimization) and MBSE (model-based systems engineering) are direct beneficiaries of the additional capabilities and collaborative approach described in this paper.
- Yes, it is free (until I die) and closed-source (until the community does not insists on making it open-source and funding sources are secured to make it happen).
- Figures make up at least half of the paper, and I had to move some code to notebooks. The requirements say: "The paper should be at most 6000 words including figures but not including references; this is about 8 pages for the published PDF", which I read as: "there is a mandatory under 6k words, but PDF is whatever it takes". I have seen papers of 18 pages from last year ([multinterp](https://proceedings.scipy.org/articles/FGCJ9164)), so I believe it should be fine. In any case, I'm open to suggestions.

nomad-vagabond committed:
Add Acknowledgments section. Fix spelling issues.

nomad-vagabond committed:
Fix issue with section references for the paper PDF building

nomad-vagabond committed:
Organize figures in a separate folder

nomad-vagabond committed:
Hotfix: update figure path for fine-tune-control-points.png

nomad-vagabond committed:
Replace paper banner

nomad-vagabond committed:
Add thumbnail.png

nomad-vagabond commented:
Paper is ready for the final review!

nomad-vagabond committed:
Add notebook with examples how to reuse curves from SplineCloud

nomad-vagabond committed:
Update notebook title

nomad-vagabond committed:
Fix plot title in the notebook with SplineCloud examples

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -455,22 +464,114 @@ The client library introduces basic capabilities that allow for fetching, recrea
+
+As described in {ref}`interactive-spline-fitting-workflow` of this paper, SplineCloud extends the capabilities of SciPy and addresses key limitations of traditional programmatic spline fitting by providing visual control over spline parameters, enabling fine-tuning through interactive manipulation of control points and knot vectors. The application of FAIR principles to regression modeling enhanc ...
+
+This is proved by the use case studies, given in {ref}`implications-and-limitations-of-splinecloud`,  that illustrate the platform’s effectiveness in addressing real-world challenges, including: complex data fitting from hydrogeological experiments; improving reusability of models based on historical data in aerodynamics; and the creation of digital libraries of empirical relations to support eng ...
```
Wrong reference? Probably Section 7

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -455,22 +464,114 @@ The client library introduces basic capabilities that allow for fetching, recrea
+
+Step-drawdown tests are performed to determine the optimal pumping rate for water boreholes by measuring how water levels respond to progressively increasing discharge rates. Traditional analysis methods assume that relationships between pumping rates and water level drop should be predictable and follow simple mathematical models, but experimental data often reveals complex non-linear behaviors  ...
+
+In [@doi:10.4314/wsa.v27i3.4978] authors developed a model for predicting drawdown behavior in step-drawdown tests based on experimental data from South Africa's Karoo formations. This model can be useful in determining sustainable pumping rates for boreholes. While relying on physical properties of fractured rock aquifers, flow dimensions, and elasticity parameters, the model still exhibits unde ...
```
Super minor detail. I feel `@doi:10.4314/wsa.v27i3.4978 developed` rendered to `Van Tonder et al., 2001 developed` reads much better

apaleyes commented on papers/vadym_pasko/main.md:
```diff
@@ -455,22 +464,114 @@ The client library introduces basic capabilities that allow for fetching, recrea
+
+[^footnote-span-efficiency-relations]: [Data relations](https://splinecloud.com/compose/dfl_omTIwGOEVQlx/relations) constructed using plot for efficiency factor [@mc-cormick]
+
+### Building Reusable Libraries of Airfoils' Performance Curves
```
Personally, I found this example to be the strongest, because it talks about what is and not what could have been. Especially because I found this reference that seems to have used this very library (consider citing here?):  https://arxiv.org/abs/2403.14979

Nevertheless, other use cases also make total sense.

nomad-vagabond replied:
@apaleyes thank you for the link to the article.  Cited. I will also add it to the reference list in my future work, related to the analysis of meta models usage in design optimization.

apaleyes reviewed (commented):
I don't have much to add to the previous review. Newly added sections are as well written as the rest of the paper, and specifically use cases do a good job in highlighting potential utility of SplineCloud. Clearly lots of proofreading works was also done, many articles and typos are fixed. Thanks @nomad-vagabond !

nomad-vagabond committed:
Incorporate comments: fix link to the Section 7, cite NN approach for airfoils performance prediction.

nomad-vagabond committed:
Fix broken section label

nomad-vagabond committed:
Update some images to take less space

nomad-vagabond commented:
@apaleyes Thanks! I implemented the latest suggestions.

nomad-vagabond commented:
Dear editors, I would like to know if I may expect new amendment requests prior to the upcoming deadline - August 9.

cdlindsey commented on papers/vadym_pasko/main.md:
```diff
@@ -0,0 +1,579 @@
+abstract: |
+  Curve fitting is a fundamental task in data science, engineering, and scientific computing, enabling researchers to extract meaningful relationships from data. However, selecting and tuning the right fitting model for complex, noisy, or multidimensional data remains a significant challenge. SciPy plays a critical role in addressing these challenges by providing robust spline fitting methods tha ...
+  
+  In order to address these challenges, we developed [SplineCloud](https://splinecloud.com/) - an open platform that provides interactive spline fitting capabilities and uses SciPy on the backend. SplineCloud allows the construction, analysis, and exchange of spline-based regression models using SciPy’s `interpolate` module. SplineCloud’s [curve fitting tool](https://splinecloud.com/features/curv ...
```
maybe "and quickly analyze the accuracy of the models" rather than "instantly analyzing the accuracy of models"

cdlindsey commented on papers/vadym_pasko/main.md:
````diff
@@ -0,0 +1,579 @@
+\frac{t_{i+k+1} - t}{t_{i+k+1} - t_{i+1}} B_{i+1,k-1}(t)
+```
+
+In {ref}`cox-de-boor` possible divisions by zero are resolved by the convention that ‘anything divided by zero is zero’ [@lyche-morken]. This formulation separates geometry (control points) from basis functions, allowing flexible manipulation and efficient computation of spline curves.
````
You might give another reference to this too. I know the spatial sign has a similar convention,

cdlindsey commented:
I made a few comments.  I have a big question too: what does all of this mean for model inference.  Are there any supports for cross validation with the adjustments to the splines?

nomad-vagabond commented:
@cdlindsey Thanks for reviewing my paper! 

Regarding your main question:

On the platform, users can compare the accuracy of different spline fits (including manually adjusted ones) by examining their RMSEs displayed in the Curves toolbox. Additional metrics such as MSE and MAE are planned for later inclusion, but are available in the code.

A formal cross-validation can also be performed in code. Given you split your data into training and test datasets and fit splines on the training dataset, then by using the client library [splinecloud-scipy](https://github.com/nomad-vagabond/splinecloud-scipy) and recreating the spline object in code (using its ID), you can use its fit_accuracy method against the test dataset:

```
col_names, test_data = load_subset(<test_subset_ID>)
RMSE = spline.fit_accuracy(test_data, method="RMSE")
```

Since spline objects also support extrapolation, cross-validation can additionally be applied to data beyond the training range. If a curve is modified on the platform, it must be reloaded in code with the load_spline function using the same curve ID before recalculating accuracy.

Cross-validation could later be implemented directly on the platform to simplify this type of analysis. Thank you for raising this important point!

Hope this answers your question.

nomad-vagabond committed:
Improve wording in the abstract
